consoles/video_stream.pm

```diff
@@ -0,0 +1,354 @@
+            if ($current_timings && $current_timings ne $self->{dv_timings}) {
+                bmwqemu::diag "Updating DV timings, new: $current_timings";
+                # yes, there is need to update DV timings, restart ffmpeg,
+                # connect_remote_video will update the timings
+                $self->disable_video;
+                $self->connect_remote_video($self->{args}->{url});
+            } elsif ($self->{dv_timings} && !$current_timings) {
+                bmwqemu::diag "video disconnected";
+                $self->disable_video;
+                $self->{dv_timings} = '';
+            }
+            $self->{dv_timings_last_check} = time;
+        }
+    }
+
+    # no video connected, don't read anything
+    return 0 unless $self->{ffmpeg};
+
+    my $have_recieved_update = 0;
+    while ($self->_receive_frame()) {
+        $have_recieved_update = 1;
+    }
+    return $have_recieved_update;
+}
+
+sub current_screen ($self) {
+    $self->update_framebuffer();
+    return unless $self->{_framebuffer};
+    return $self->{_framebuffer};
+}
+
+sub request_screen_update ($self, @) {
+    if (!$self->update_framebuffer()) {
+        # check if it isn't stalled, perhaps we missed resolution change?
+        my $time_since_last_update = time - $self->{_last_update_received};
+        if ($self->{ffmpeg} && $time_since_last_update > STALL_THRESHOLD) {
+            # reconnect, it will refresh the device settings too
+            $self->disable_video;
+            $self->connect_remote_video($self->{args}->{url});
+        }
+    }
+}
+
+sub type_string ($self, $args) {
+    my $seconds_per_keypress = 1 / (get_var('STREAM_TYPING_LIMIT', STREAM_TYPING_LIMIT_DEFAULT) || 1);
+
+    # further slow down if being asked for.
+    # 250 = magic default from testapi.pm
+
+    # Note: the intended use of max_interval is the bootloader. The bootloader
+    # prompt drops characters when typing quickly. This problem mostly occurs
+    # in the bootloader. Humans notice because they look at the screen while
+    # typing. So this loop should be replaced by some true 'looking at the
+    # screen while typing', e.g. waiting for no more screen updates 'in the
+    # right area'.  For now, just waiting is good enough: The slow-down only
+    # affects the bootloader sequence.
+    if (($args->{max_interval} // 250) < 250) {
+        # according to 	  git grep "type_string.*, *[0-9]"  on
+        #   https://github.com/os-autoinst/os-autoinst-distri-opensuse,
+        # typical max_interval values are
+        #   4ish:  veeery slow
+        #   15ish: slow
+        $seconds_per_keypress = $seconds_per_keypress + 1 / sqrt($args->{max_interval});
+    }
+
+    for my $letter (split("", $args->{text})) {
+        next if ($letter eq "\r");
+        my $charmap = {
+            "\t" => 'tab',
+            "\n" => 'ret',
+            "\b" => 'backspace',
+            "\e" => 'esc',
+            " " => 'space',
+        };
```

Comment:
Yes, it would be more efficient to define this at the outermost scope. I would put it like this at the top of the file:
```perl
my $CHARMAP = {
            "\t" => 'tab',
            "\n" => 'ret',
            "\b" => 'backspace',
            "\e" => 'esc',
            " " => 'space',
        };
```